src/_pytest/capture.py

```diff
@@ -497,345 +584,319 @@ def readouterr(self) -> CaptureResult:
+            return "fixture %s" % self._capture_fixture.request.fixturename
+        return False
 
-class FDCapture(FDCaptureBinary):
-    """Capture IO to/from a given os-level filedescriptor.
+    # Global capturing control
 
-    snap() produces text
-    """
+    def is_globally_capturing(self):
+        return self._method != "no"
 
-    # Ignore type because it doesn't match the type in the superclass (bytes).
-    EMPTY_BUFFER = ""  # type: ignore
+    def start_global_capturing(self):
+        assert self._global_capturing is None
+        self._global_capturing = _get_multicapture(self._method)
+        self._global_capturing.start_capturing()
 
-    def snap(self):
-        self.tmpfile.seek(0)
-        res = self.tmpfile.read()
-        self.tmpfile.seek(0)
-        self.tmpfile.truncate()
-        return res
+    def stop_global_capturing(self):
+        if self._global_capturing is not None:
+            self._global_capturing.pop_outerr_to_orig()
+            self._global_capturing.stop_capturing()
+            self._global_capturing = None
 
-    def writeorg(self, data):
-        """ write to original file descriptor. """
-        super().writeorg(data.encode("utf-8"))  # XXX use encoding of original stream
+    def resume_global_capture(self):
+        # During teardown of the python process, and on rare occasions, capture
+        # attributes can be `None` while trying to resume global capture.
+        if self._global_capturing is not None:
+            self._global_capturing.resume_capturing()
 
+    def suspend_global_capture(self, in_=False):
+        if self._global_capturing is not None:
+            self._global_capturing.suspend_capturing(in_=in_)
 
-class SysCaptureBinary:
+    def suspend(self, in_=False):
+        # Need to undo local capsys-et-al if it exists before disabling global capture.
+        self.suspend_fixture()
+        self.suspend_global_capture(in_)
 
-    EMPTY_BUFFER = b""
-    _state = None
+    def resume(self):
+        self.resume_global_capture()
+        self.resume_fixture()
 
-    def __init__(self, fd, tmpfile=None):
-        name = patchsysdict[fd]
-        self._old = getattr(sys, name)
-        self.name = name
-        if tmpfile is None:
-            if name == "stdin":
-                tmpfile = DontReadFromInput()
-            else:
-                tmpfile = CaptureIO()
-        self.tmpfile = tmpfile
+    def read_global_capture(self):
+        return self._global_capturing.readouterr()
 
-    def __repr__(self):
-        return "<{} {} _old={} _state={!r} tmpfile={!r}>".format(
-            self.__class__.__name__,
-            self.name,
-            hasattr(self, "_old") and repr(self._old) or "<UNSET>",
-            self._state,
-            self.tmpfile,
-        )
+    # Fixture Control
 
-    def start(self):
-        setattr(sys, self.name, self.tmpfile)
-        self._state = "started"
+    def set_fixture(self, capture_fixture: "CaptureFixture") -> None:
```

Comment:
at first glance i believe that pair might yield nicer code if it was a context manager instead

Reply:
So, it actually was a context manager, but the commits removes it. The reason is that the context manager adds some indirection which makes it more difficult to refactor further (in later PRs).